doc/storagemond/storagemond-hld.md

```diff
@@ -0,0 +1,377 @@
+
+
+#### **2.4.4 storagemond Class Diagram**
+
+![image.png](images/StoragemonDaemonClassDiagram.png)
+
+## **3. StateDB Schema**
+```
+; Defines information for each Storage Disk in a device
+
+key                 = STORAGE_INFO|<disk_name>  ; This key is for information about a specific storage disk - STORAGE_INFO|SDX
+
+; field             = value
+
+device_model        = STRING                    ; Describes the Vendor information of the disk                                           (Static)
+serial              = STRING                    ; Describes the Serial number of the disk                                                (Static)
+temperature_celsius = STRING                    ; Describes the operating temperature of the disk in Celsius                             (Dynamic)
+fs_io_reads         = STRING                    ; Describes the total number of filesystem reads completed successfully                  (Dynamic)
+fs_io_writes        = STRING                    ; Describes the total number of filesystem writes completed successfully                 (Dynamic)
+disk_io_reads       = STRING                    ; Describes the total number of reads completed successfully from the SSD (LBAs read)    (Dynamic)
+disk_io_writes      = STRING                    ; Describes the total number of writes completed on the SSD (LBAs written)               (Dynamic)
+reserved_blocks     = STRING                    ; Describes the reserved blocks count of the SSD                                         (Dynamic)
+firmware            = STRING                    ; Describes the Firmware version of the SSD                                              (Dynamic)
+health              = STRING                    ; Describes the overall health of the SSD as a % value based on several SMART attrs      (Dynamic)
+```
+
+NOTE: 'LBA' stands for Logical Block Address. To get the raw value in bytes, multiply by the disk's logical block address sze (typically 512 bytes).<br>
+
+Example: For an SSD with name 'sda', the STATE_DB entry would be:
+
+```
+root@str2-7050cx3-acs-01:~# docker exec -it database bash
+root@str2-7050cx3-acs-01:/# redis-cli -n 6
+127.0.0.1:6379[6]> keys STORAGE*
+1) "STORAGE_INFO|mmcblk0"
+2) "STORAGE_INFO|sda"
+127.0.0.1:6379[6]>
+127.0.0.1:6379[6]> hgetall STORAGE_INFO|sda
+ 1) "device_model"
+ 2) "SATA SSD"
+ 3) "serial"
+ 4) "SPG2043056Z"
+ 5) "firmware"
+ 6) "FW1241"
+ 7) "health"
+ 8) "N/A"
+ 9) "temperature"
+10) "30"
+11) "fs_io_reads"
+12) "28753"
+13) "fs_io_writes"
+14) "92603"
+15) "disk_io_reads"
+16) "15388141951"
+17) "disk_io_writes"
+18) "46070618960"
+19) "reserved_blocks"
+20) "32"
+
+```
+
+## Future Work
+
+1. Full support for eMMC
+2. Support for USB and NVMe storage disks
+3. Refactor `ssdutil` [in sonic-utilities](https://github.com/sonic-net/sonic-utilities/tree/master/ssdutil) to cover all storage types, including changing the name of the utility to 'storageutil'
```

Comment:
As discussed please try to introduce a separate CLI in the future to access DB and get output. We need to preserve ssdutil behavior to access ssd in case pmon or database is unavailable.

Reply:
A sonic utility [ssdutil](https://github.com/sonic-net/sonic-utilities/tree/master/ssdutil) currently exists that has limited capability. We would be enhancing this utility as part of the current effort.